File: Algorithms/ValidAnagram/validAnagram.cc
/*
copyright xu(xusiwei1236@163.com), all right reserved.

Valid Anagram
==============

Given two strings s and t, write a function to determine if t is an anagram of s. 

For example,
s = "anagram", t = "nagaram", return true.
s = "rat", t = "car", return false.


Note:
You may assume the string contains only lowercase alphabets.

Follow up:
What if the inputs contain unicode characters? How would you adapt your solution to such case?
*/

#include <string.h>
#include <string>
#include <iostream>
using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t) {
        const int N = 256;
        char map1[N] = {0}, map2[N] = {0};

        for (int i = 0; i < s.length(); i++) {
            map1[s[i]]++;
        }
        for (int i = 0; i < t.length(); i++) {
            map2[t[i]]++;
        }
        return memcmp(map1, map2, sizeof(map1)) == 0;
    }

};

int main(int argc, char* argv[])
{
    string s = "anagram";
    string t = "nagaram";

    if (argc > 2) {
        s = argv[1];
        t = argv[2];
    }
    cout << s << endl;
    cout << t << endl;
    cout << Solution().isAnagram(s, t) << endl;
    return 0;
}

File: Algorithms/ValidAnagram/validAnagram.py
#!/usr/bin/env python
'''
copyright xu(xusiwei1236@163.com), all right reserved.

Valid Anagram
==============

Given two strings s and t, write a function to determine if t is an anagram of s. 

For example,
s = "anagram", t = "nagaram", return true.
s = "rat", t = "car", return false.


Note:
You may assume the string contains only lowercase alphabets.

Follow up:
What if the inputs contain unicode characters? How would you adapt your solution to such case?
'''
import sys

class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        cs = [0] * 256
        ct = [0] * 256
        for c in s: cs[ord(c)] += 1
        for c in t: ct[ord(c)] += 1
        return cs == ct

if __name__ == '__main__':
    s = 'anagram'
    t = 'nagaram'

    if len(sys.argv) > 1: s = sys.argv[1]
    if len(sys.argv) > 2: t = sys.argv[2]

    print s, t
    print Solution().isAnagram(s, t)
